core/src/actions_normalize.cpp

```diff
@@ -160,7 +160,13 @@ bool km::core::actions_normalize(
     if(app_context_nfd.compare(cached_context_string) == 0) {
       break;
     }
-    app_context_string.remove(app_context_string.length()-1);
+
+    // remove the last UChar32
+    int32_t lastUChar32 = app_context_string.length()-1;
+    // adjust pointer to get the entire char (i.e. so we don't slice a non-BMP char)
+    lastUChar32 = app_context_string.getChar32Start(lastUChar32);
+    // remove the UChar32 (1 or 2 code units)
+    app_context_string.remove(lastUChar32);
```

Comment:
Does this handle split surrogate pairs at start of context? Do we need a unit test for this?

Reply:
I think if there were a mismatched trail it would be removed singly. Can add a test. That situation would not be regressed by this, though, because the first trail bike would be considered a single code unit under both the old and the new calculation

Reply:
In other words, I would recommend that we consider merging this… I think a mismatch surrogate at the beginning of context should be handled separately, perhaps in a different place than this function

Reply:
agree we can consider split surrogate situation separately

Reply:
I verified in the debugger that bmp characters work properly here, too, by the way